### PR TITLE
[JIT] Use byteCount when deciding what instruction sequence to use in genCodeForCpBlkUnroll

### LINKED ISSUE #64125: Assertion failed 'helper.CanEncodeAllOffsets(REGSIZE_BYTES)' during 'Generate code'
```c#
// Found by Antigen

using System;
using System.Collections.Generic;
using System.Runtime.CompilerServices;
public class TestClass
{
    public struct S1
    {
        public struct S1_D1_F2
        {
            public uint uint_1;
        }
        public struct S1_D1_F4
        {
            public struct S1_D2_F1
            {
                public S1.S1_D1_F2 s1_s1_d1_f2_3;
                public short short_4;
            }
            public struct S1_D2_F2
            {
                public decimal decimal_5;
            }
        }
        public ulong ulong_0;
        public int int_2;
    }
    public struct S2
    {
        public struct S2_D1_F1
        {
            public struct S2_D2_F2
            {
                public long long_8;
                public S1 s1_9;
                public S1.S1_D1_F2 s1_s1_d1_f2_10;
            }
            public ulong ulong_7;
            public S1.S1_D1_F4 s1_s1_d1_f4_11;
        }
    }
    public struct S3
    {
        public struct S3_D1_F1
        {
            public struct S3_D2_F1
            {
                public decimal decimal_12;
            }
        }
        public struct S3_D1_F2
        {
            public S1.S1_D1_F4.S1_D2_F1 s1_s1_d1_f4_s1_d2_f1_13;
            public uint uint_15;
            public S3.S3_D1_F1.S3_D2_F1 s3_s3_d1_f1_s3_d2_f1_16;
        }
    }
    public struct S4
    {
        public struct S4_D1_F1
        {
            public struct S4_D2_F1
            {
                public ushort ushort_18;
                public S1.S1_D1_F4.S1_D2_F1 s1_s1_d1_f4_s1_d2_f1_19;
                public S2 s2_20;
            }
            public S3.S3_D1_F2 s3_s3_d1_f2_21;
        }
    }
    public struct S5
    {
        public string string_22;
    }
    static bool s_bool_24 = false;
    static byte s_byte_25 = 5;
    static char s_char_26 = '9';
    static S1.S1_D1_F4.S1_D2_F2 s_s1_s1_d1_f4_s1_d2_f2_40 = new S1.S1_D1_F4.S1_D2_F2();
    static S3.S3_D1_F2 s_s3_s3_d1_f2_48 = new S3.S3_D1_F2();
    static S4.S4_D1_F1.S4_D2_F1 s_s4_s4_d1_f1_s4_d2_f1_50 = new S4.S4_D1_F1.S4_D2_F1();
    static S5 s_s5_55 = new S5();
    byte byte_57 = 5;
    char char_58 = 'S';
    decimal decimal_59 = -4.9875m;
    double double_60 = -4.954022988505747;
    int int_62 = 0;
    long long_63 = 1;
    float float_65 = 0.021052632f;
    string string_66 = "1K1GWJ";
    ushort ushort_67 = 2;
    uint uint_68 = 51;
    ulong ulong_69 = 5;
    S1.S1_D1_F2 s1_s1_d1_f2_70 = new S1.S1_D1_F2();
    S1.S1_D1_F4.S1_D2_F1 s1_s1_d1_f4_s1_d2_f1_71 = new S1.S1_D1_F4.S1_D2_F1();
    S1.S1_D1_F4.S1_D2_F2 s1_s1_d1_f4_s1_d2_f2_72 = new S1.S1_D1_F4.S1_D2_F2();
    S1.S1_D1_F4 s1_s1_d1_f4_73 = new S1.S1_D1_F4();
    S1 s1_74 = new S1();
    S2.S2_D1_F1.S2_D2_F2 s2_s2_d1_f1_s2_d2_f2_75 = new S2.S2_D1_F1.S2_D2_F2();
    S2.S2_D1_F1 s2_s2_d1_f1_76 = new S2.S2_D1_F1();
    S2 s2_77 = new S2();
    S3.S3_D1_F1.S3_D2_F1 s3_s3_d1_f1_s3_d2_f1_78 = new S3.S3_D1_F1.S3_D2_F1();
    S3.S3_D1_F1 s3_s3_d1_f1_79 = new S3.S3_D1_F1();
    S3.S3_D1_F2 s3_s3_d1_f2_80 = new S3.S3_D1_F2();
    S3 s3_81 = new S3();
    S4.S4_D1_F1.S4_D2_F1 s4_s4_d1_f1_s4_d2_f1_82 = new S4.S4_D1_F1.S4_D2_F1();
    S4.S4_D1_F1 s4_s4_d1_f1_83 = new S4.S4_D1_F1();
    S5 s5_87 = new S5();
    [MethodImpl(MethodImplOptions.NoInlining)]
    public decimal LeafMethod3()
    {
        unchecked
        {
            return 15+4;
        }
    }
    public string LeafMethod10()
    {
        unchecked
        {
            return string_66 = s5_87.string_22 = 15+4+ s_s5_55.string_22;
        }
    }
    public S4.S4_D1_F1.S4_D2_F1 Method2(S3 p_s3_197)
    {
        unchecked
        {
            return s_s4_s4_d1_f1_s4_d2_f1_50;
        }
    }
    public void Method0()
    {
        unchecked
        {
            char char_317 = 'G';
            string string_325 = "AFBKJAY";
            S3.S3_D1_F2 s3_s3_d1_f2_345 = new S3.S3_D1_F2();
            S5 s5_355 = new S5();
            if (s_byte_25 <= (byte_57 %= 15+4))
            {
                switch (char_317 = s_char_26 = char_58 = s_char_26)
                {
                    case 'B':
                    {
                        if (s_bool_24)
                        {
                            switch (15+4+ string_66+ (s5_355.string_22 = "R8FVKBU6" + string_325)+ (string_325 = LeafMethod10() + s5_355.string_22))
                            {
                                case "":
                                {
                                    break;
                                }
                                default:
                                {
                                    break;
                                }
                            }
                        }
                        else
                        {
                            s3_s3_d1_f2_345.s3_s3_d1_f1_s3_d2_f1_16.decimal_12 %= 15*4+ s_s1_s1_d1_f4_s1_d2_f2_40.decimal_5- LeafMethod3()+ (s_s3_s3_d1_f2_48.s3_s3_d1_f1_s3_d2_f1_16.decimal_12 %= s3_s3_d1_f2_80.s3_s3_d1_f1_s3_d2_f1_16.decimal_12 + LeafMethod3()+ 20)/ 1.0303030303030303030303030303m + LeafMethod3()- s4_s4_d1_f1_83.s3_s3_d1_f2_21.s3_s3_d1_f1_s3_d2_f1_16.decimal_12+ 59+ 99;
                        }
                        break;
                    }
                    case 'N':
                    {
                        break;
                    }
                    default:
                    {
                        break;
                    }
                }
            }
            else
            {
            }
            s4_s4_d1_f1_s4_d2_f1_82 = Method2(s3_81);
            return;
        }
    }
    public static void Main(string[] args)
    {
        new TestClass().Method0();
    }
}

/*
Environment:

set COMPlus_JITMinOpts=1
set COMPlus_TieredCompilation=1
set COMPlus_AltJitName=clrjit_universal_arm64_x64.dll
set COMPlus_AltJit=Method0

Assert failure(PID 57256 [0x0000dfa8], Thread: 28392 [0x6ee8]): Assertion failed 'helper.CanEncodeAllOffsets(REGSIZE_BYTES)' in 'TestClass:Method0():this' during 'Generate code' (IL size 435)
    File: D:\git\dotnet-runtime\src\coreclr\jit\codegenarmarch.cpp Line: 2908
    Image: d:\git\dotnet-runtime\artifacts\tests\coreclr\windows.x64.Checked\tests\Core_Root\CoreRun.exe
*/
```

Repro can be shorter; I didn't try reducing it.

### PR DESCRIPTION
**Pull request by echesakov**
Here is what happens in #64125:

The CodeGen runs https://github.com/dotnet/runtime/blob/ea3881b8423d18f7bb1185afde67f848d64d398e/src/coreclr/jit/codegenarmarch.cpp#L2690
where `srcOffset` = `512`, `dstOffset` = `252`, `size` = `16`:

When the JIT validates whether we can encode all stores and loads at https://github.com/dotnet/runtime/blob/ea3881b8423d18f7bb1185afde67f848d64d398e/src/coreclr/jit/codegenarmarch.cpp#L2801 
`canEncodeAllLoads` is `true` and `canEncodeAllStores` is `false` since internally in the helper at https://github.com/dotnet/runtime/blob/ea3881b8423d18f7bb1185afde67f848d64d398e/src/coreclr/jit/codegenarmarch.cpp#L2419 we take a path of using a sequence of `ldr`/`str` instructions and for these the offsets need to be `8*k`.

The next step is the JIT updates the `dstOffset` to new value of `512`, since the JIT knows that `ldr` instructions were encodable and, we can use the same starting offset to make `str` instruction encodable as well.

However, when we run an assert at https://github.com/dotnet/runtime/blob/ea3881b8423d18f7bb1185afde67f848d64d398e/src/coreclr/jit/codegenarmarch.cpp#L2907 we take another path in the helper of using a sequence of `ldp`/`stp` instructions, but such sequence can not be encoded, hence, the assert fails.

The flaw in the algorithm is that the decision whether to use `ldp`/`stp` vs. `ldr`/`str` cannot be altered once the JIT made an assumption about `canEncodeAllLoads` and `canEncodeAllStores` values. Hence, the decision needs to be invariant to the value of `dstOffset` (and `dstOffsetAligned`).

### CODE REVIEW

**Comment by echesakovMSFT** [comment]
The only tests that are failed Mono/Wasm. Maybe the "unsafe" magic I am doing in the test violate ECMA-335 rules? Or there is an issue with there runtimes

File: src/tests/JIT/Regression/JitBlue/Runtime_64125/Runtime_64125.tt [hunk 1]
```diff
@@ -0,0 +1,90 @@
+﻿<#@ template debug="false" hostspecific="true" language="C#" #>
+<#@ assembly name="System.Core" #>
+<#@ import namespace="System.Linq" #>
+<#@ output extension=".cs" #>
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
```
Comment (BruceForstall) [inline src/tests/JIT/Regression/JitBlue/Runtime_64125/Runtime_64125.tt:7]: Looks like our standard file header doesn't include this "See the ..." line anymore.

File: src/tests/JIT/Regression/JitBlue/Runtime_64125/Runtime_64125.tt [hunk 2]
```diff
@@ -0,0 +1,90 @@
+﻿<#@ template debug="false" hostspecific="true" language="C#" #>
```
Comment (BruceForstall) [inline src/tests/JIT/Regression/JitBlue/Runtime_64125/Runtime_64125.tt:1]: Can you include comments/instructions here describing what this ".tt" file is, and more importantly, how to regenerate the .cs file if changes are made. Also, add a comment into the generated .cs file that it is a generated file from this .tt file and don't modify it directly.

File: src/tests/JIT/Regression/JitBlue/Runtime_64125/Runtime_64125.tt [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    {
+<#
+    foreach (int byteCount in byteCounts)
+    foreach (int srcOffset in offsets)
+    foreach (int dstOffset in offsets)
+    {
+        string structByteCount = $"Struct{byteCount}Bytes";
+#>
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        public unsafe void Copy<#= byteCount #>BytesFromLocationAtOffset<#= srcOffset #>ToLocationAtOffset<#= dstOffset #>(AnyOffset* dst, AnyOffset* src)
```
Comment (BruceForstall) [inline src/tests/JIT/Regression/JitBlue/Runtime_64125/Runtime_64125.tt:57]: Presumably you could write a simple byte-by-byte comparison function that would verify the bytes were actually copied (you'd need to initialize the src and dst to something different, then recognize the copied pattern correctly). Then you would be checking that the right bytes actually did get copied, not just that we didn't crash.

File: src/coreclr/jit/codegenarmarch.cpp [hunk 4]
```diff
@@ -2416,7 +2416,7 @@ class CopyBlockUnrollHelper
 
         const int dstOffsetAligned = AlignUp((UINT)dstOffset, storePairRegsAlignment);
 
-        if (endDstOffset - dstOffsetAligned >= storePairRegsWritesBytes)
+        if (byteCount >= (unsigned)storePairRegsWritesBytes)
```
Comment (BruceForstall) [inline src/coreclr/jit/codegenarmarch.cpp:2419]: `UnrollInitBlock` has somewhat similar logic. Does it need to be updated?

**Review by BruceForstall — APPROVED** [review]
LGTM, assuming the tests pass. Maybe you disable them for mono/wasm to start with?

File: src/coreclr/jit/codegenarmarch.cpp (hunk 4 above)
Comment (echesakovMSFT) [reply src/coreclr/jit/codegenarmarch.cpp:2419]: No - there is only one `dstOffset` in InitBlock case. 

If a store instruction couldn't be encoded during initial `TryEncodeAllOffsets` call - we spill the whole `dstAddr` to a register (effectively, setting `dstOffset` to `0`). In other words, there is only one *"degree of freedom"* for InitBlock while for CopyBlock we have two (and in this case, changing value of one `dstOffset` caused going from a state `{canEncodeAllLoads: true, canEncodeAllStores: false}` to `{canEncodeAllLoads: false, canEncodeAllStores: true}`)

File: src/tests/JIT/Regression/JitBlue/Runtime_64125/Runtime_64125.tt (hunk 1 above)
Comment (echesakovMSFT) [reply src/tests/JIT/Regression/JitBlue/Runtime_64125/Runtime_64125.tt:7]: Okay, copied this from a wrong place

File: src/tests/JIT/Regression/JitBlue/Runtime_64125/Runtime_64125.tt (hunk 3 above)
Comment (echesakovMSFT) [reply src/tests/JIT/Regression/JitBlue/Runtime_64125/Runtime_64125.tt:60]: Updated, added the validation.

File: src/tests/JIT/Regression/JitBlue/Runtime_64125/Runtime_64125.tt (hunk 2 above)
Comment (echesakovMSFT) [reply src/tests/JIT/Regression/JitBlue/Runtime_64125/Runtime_64125.tt:1]: Updated.

File: src/tests/JIT/Regression/JitBlue/Runtime_64125/Runtime_64125.tt (hunk 1 above)
Comment (echesakovMSFT) [reply src/tests/JIT/Regression/JitBlue/Runtime_64125/Runtime_64125.tt]: Updated.

**Comment by echesakovMSFT** [comment]
> Maybe you disable them for mono/wasm to start with?

Tried disabling the tests with issues.targets. See if I've done this right

**Review by BruceForstall — APPROVED** [review]

**Pull request merged**